### ISSUE TITLE
Suggestion: Have an 'end date'

### ISSUE THREAD

**Issue by Ipstenu** [humanmade/repeatable-posts#7]
Current this runs until the heat death of the universe. It would be nice to have the ability to stop repeating on a date or after X times.

**Comment by noplanman** [comment]
I think this is a great idea! Saves from remembering to edit the original post when you don't want it to be repeated any more.

@willmot You ok with me starting to implement this?

**Comment by willmot** [comment]
Absolutely!

**Comment by noplanman** [comment]
@willmot I thought of changing the post meta field to an array in this format (values are just examples obviously):
`array( 'schedule' => 'weekly', 'ending' => '2016-01-01 12:00:00' );`
or
`array( 'schedule' => 'monthly', 'ending' => '3' );`

The first example would be a repeating post until the defined date.
The second one would be repeated the chosen number of times. This value would be decreased by 1 after every successful publish.

What do you think? I thought the word `ending` could be used for both scenarios, but maybe there is a more fitting one?

**Comment by willmot** [comment]
This looks sensible 👍 I'd probably use separate keys for the different options, maybe `end_date` and `count`?

**Comment by noplanman** [comment]
I was thinking about that too, but what if both keys are set by directly editing the post meta in the database? Which one would have the *power*?
My idea was to keep it simple and foolproof somehow. If there is a date, post until that date is reached; if it's a number, post and decrement until it's 0.

The other option would be to manage it by date only but offer both options on the front end.
i.e. You can either choose `end_date` or `count` and in the back end it does the calculation and only saves the date for example (or the count, whichever).

**Comment by willmot** [comment]
> My idea was to keep it simple and foolproof somehow. If there is a date, post until that date is reached; if it's a number, post and decrement until it's 0.

This 👍, basically whichever one is hit first.

> The other option would be to manage it by date only but offer both options on the front end.
i.e. You can either choose end_date or count and in the back end it does the calculation and only saves the date for example (or the count, whichever).

That's a pretty smart idea. Happy with either way tbh

**Comment by Ipstenu** [comment]
Whichever hits first seems sensible.

Repeat weekly 6 times, or until October 31st, 2015.

Then if I'm only 4 times by the 31st, it still stops :)

**Comment by noplanman** [comment]
@Ipstenu You're right, it seems like the better way to go.
Also makes it easier to define repeatable posts :blush:

What do you think of these mockups?
![repeatable-post-meta-selected](https://cloud.githubusercontent.com/assets/9423417/10434459/f21be086-7119-11e5-8531-87b0f5fc8083.png)
*selected*

![repeatable-post-meta-edit](https://cloud.githubusercontent.com/assets/9423417/10434458/f21a5c7a-7119-11e5-9113-1b6b7ea2200b.png)
*editing*

I thought of going with the existing date selection method that WP uses, so it's a bit big, I know. Also, if one would like to only repeat the post, say 100x (without having the date as a second condition), the date would have to be set to a very late future, to make sure that the date isn't the limiting factor.

Any ideas on how to make a sensible input for this?

**Comment by noplanman** [comment]
I've given this some more thought and have come up with something like this:
![repeatable-posts-count-end_time](https://cloud.githubusercontent.com/assets/9423417/10717741/a09c0e1e-7b61-11e5-89c6-669bd37f0ce5.png)

This way, both options can be set and activated independently, or used together.
Adding some JavaScript to disable the inputs when the checkboxes aren't checked (enabling them when checked) will make it feel better. Also, if **No** is selected, the inputs aren't displayed at all.

What do you think?

**Comment by willmot** [comment]
I think this is slick. Although maybe those should be radio buttons instead of checkboxes, to make it obvious you choose one.

**Comment by noplanman** [comment]
@willmot I deliberately made them checkboxes, to incorporate what @Ipstenu suggested:
> Whichever hits first seems sensible.
> Repeat weekly 6 times, or until October 31st, 2015.
> Then if I'm only 4 times by the 31st, it still stops :)

The other advantage of having checkboxes (when neither is checked), is to keep the initial idea of the plugin alive, to repeat posts ***until the heat death of the universe*** :wink:

**Comment by willmot** [comment]
I assumed that @Ipstenu was just suggested a method of solving the issue of having both set at once. We can avoid that issue in most cases by just having radio buttons. I don't think there's a legitimate use-case for having both set?

**Comment by noplanman** [comment]
In that case I misunderstood.
So if somebody wants to repeat indefinitely, they just set a date way in the future.

:radio_button:  Radio buttons it is then :+1:

**Comment by willmot** [comment]
@noplanman if they want it to run indefinately then they wouldn't set either I guess

**Comment by noplanman** [comment]
Which would then require an additional radio button or something to "unset" it, if it was set at some point.
The alternative would be to set repeating to **No**, then Update the post, then set the schedule again.

Radio buttons, once set, normally can't be unset. I'm just thinking of a way to make it intuitive and easier for users to understand.

With Javascript we could easily "deselect" a selected radio when clicked, making them both unselected, but for users with JS deactivated the problem persists.

**Comment by willmot** [comment]
Yeah that's true, I don't like the idea of a non-standard JS solution to that. I'm not sure of the best way forward though, perhaps a 3rd radio button which is selected by default `no limit`, however I worry we're getting to complex and clunky with the UI with 3 radio buttons.

Perhaps we could just have a single checkbox

- [ ] Limit, Times `[====]` Or, Until `[====]`

**Comment by noplanman** [comment]
So back to the initial idea that @Ipstenu proposed?
Maybe we could add a 3rd radio button "**Forever**" right next to the schedule dropdown? At least that would keep it small-ish in size.

I totally agree with you on keeping it as simple and sleek as possible, but at the same time we should ask ourselves how often this option gets set. Normally just once per new post, if it is to be repeated. So not that often.

**Comment by willmot** [comment]
@noplanman  yeah I think that works, I had a little play around, how about something like this:

![screen shot 2015-10-30 at 12 12 16](https://cloud.githubusercontent.com/assets/308507/10845409/803c0c90-7eff-11e5-824b-443c574252f5.png)

Obviously those options would only show when "edit" was clicked and something other than "No" was selected.

**Comment by noplanman** [comment]
Yep, that works. I think it needs an "**Until:**" in there somewhere, otherwise that date input is confusing. It's not clear what the date is being set for.
Maybe not show the time like in my example above?

**Comment by willmot** [comment]
@noplanman yeah good point on both unit and not showing time 👍

**Comment by noplanman** [comment]
@willmot Sorry for the epic delay, taking this on again.

Regarding the post meta option to save:
I'd prefer to have a single option per post, saving all information as an array. But adding a new option which just saves the repetition end is much easier to implement, as the current structure can stay, it's just being extended. Making an array-option means changing most existing code to accommodate it.

What's your view on this?

I've managed to get the JS part to work like WordPress does it with the dynamic options, so that's good :blush:
Now it's just down to finalising the thoughts of how to implement the option saving.

**Comment by willmot** [comment]
@noplanman awesome!

It's a been a while so I'm not really sure re. the option saving, maybe open a PR with what you have and we can take it from there?

**Comment by noplanman** [comment]
Ok, will do! :+1: 

Sorry again for the delay.

**Comment by noplanman** [comment]
Damn, this is more complex than I had anticipated, will need a bit more time...

Should the "repeat" posts still have a special post status even after the repeating has stopped?
The way I initially thought of it, was to completely remove the "repeating-post" meta data from the posts that have finished repeating, but that would mean the status would be gone too, unless we introduce a new seperate meta field to mark posts as "having been repeat(ing) ones", which I really don't think is necessary.

**Comment by willmot** [comment]
I'd leave it with the same status but just dynamically change the messaging in the admin to say "repeating finished" or something.

**Comment by noplanman** [comment]
Hi @willmot 
I'm really busy with other projects right now and can't find the time and motivation to work on this at the moment. I've committed the current progress to my branch but it's still incomplete, just in case somebody wants to continue working on it or use it as a springboard to implement this feature.

Shall I make a PR of the code I have so far for you and others to work on?

You can find my branch here: https://github.com/noplanman/repeatable-posts/tree/7-repetition_end

**Comment by willmot** [comment]
Hey @noplanman no worries and thanks for the heads-up and your work so far! If you could PR your progress that would be great 👍
